contracts/dtoken/src/lib.rs

```diff
@@ -36,14 +30,103 @@ impl Default for Dtoken {
     }
 }
 
+#[ext_contract(weth_token)]
+trait WethTokenInterface {
+    fn internal_deposit_with_registration(&mut self, account_id: AccountId, amount: Balance);
+    fn internal_withdraw_with_registration(&mut self, account_id: AccountId, amount: Balance);
+}
+
+const WETH_TOKEN_ACCOUNT_ID: &str = "dev-1639664095868-76379577854961";
+const NO_DEPOSIT: Balance = 0;
+const BASE_GAS: Gas = 80_000_000_000_000;
+
 #[near_bindgen]
 impl Dtoken {
-    pub fn supply(amount: Balance) {
-        //TODO: supply implementation
-    }
-
-    pub fn withdraw(amount: Balance) {
-        //TODO: withdraw implementation
+    pub fn supply(&mut self, amount: Balance) {
+        let dtoken_account_id = env::current_account_id();
+        let predecessor_account_id = env::predecessor_account_id();
+
+        log!("dtoken_account_id: {}", dtoken_account_id);
+        log!("signer_account_id: {}", predecessor_account_id);
+
+        let weth_token_account_id: AccountId =
+            AccountId::try_from(WETH_TOKEN_ACCOUNT_ID.clone().to_string()).unwrap();
+
+        //mock, predecessor_account_id have weth_tokens
+        weth_token::internal_deposit_with_registration(
```

Comment:
@TymofiiSavinov this should be transfer underlying (WETH) from user to dtoken and mint dtokens to user